Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,90 @@
+`n : ℕ`, `h(n)` is equal to the coefficient of `Xⁿ` in the power series expansion of `p/(1 - X)ᵈ`.
+This `h` is unique and is called the Hilbert polynomial of `p` and `d` (`Polynomial.hilbert p d`).
+
+## Main definitions
+
+* `Polynomial.hilbert p d`. If `F` is a field with characteristic `0`, `p : F[X]` and `d : ℕ`, then
+  `Polynomial.hilbert p d : F[X]` is the polynomial whose value at `n` equals the coefficient of
+  `Xⁿ` in the power series expansion of `p/(1 - X)ᵈ`
+
+## TODO
+
+* Prove that `Polynomial.hilbert p d : F[X]` is the polynomial whose value at `n` equals the
+  coefficient of `Xⁿ` in the power series expansion of `p/(1 - X)ᵈ`
+
+* Hilbert polynomials of graded modules.
+-/
+
+open BigOperators Nat PowerSeries
+
+namespace Polynomial
+
+section greatestFactorOneSubNotDvd
+
+variable {R : Type*} [CommRing R] (p : R[X]) (hp : p ≠ 0) (d : ℕ)
+
+/--
+Given a polynomial `p`, the factor `f` of `p` such that the product of `f` and
+`(1 - X : R[X]) ^ p.rootMultiplicity 1` equals `p`. We define this here because if `p` is divisible
+by `1 - X`, then the expression `p/(1 - X)ᵈ` can be reduced. We want to construct the Hilbert
+polynomial based on the most reduced form of the fraction `p/(1 - X)ᵈ`. Later we will see that this
+method of construction makes it much easier to calculate the specific degree of the Hilbert
+polynomial.
+-/
+noncomputable def greatestFactorOneSubNotDvd : R[X] :=
+  ((- 1 : R[X]) ^ p.rootMultiplicity 1) *
+  (exists_eq_pow_rootMultiplicity_mul_and_not_dvd p hp 1).choose
+
+end greatestFactorOneSubNotDvd
+
+variable (F : Type*) [Field F] [CharZero F]
+
+/--
+A polynomial which makes it easier to define the Hilbert polynomial. See also the theorem
+`Polynomial.preHilbert_eq_choose_sub_add`, which states that for any `d k n : ℕ` with `k ≤ n`,
+`(Polynomial.preHilbert d k).eval (n : F) = (n - k + d).choose d`.
+-/
+noncomputable def preHilbert (d k : ℕ) : F[X] :=
+  (d.factorial : F)⁻¹ • ((ascPochhammer F d).comp (X - (C (k : F)) + 1))
+
+theorem preHilbert_eq_choose_sub_add (d k n : ℕ) (hkn : k ≤ n):
+    (preHilbert F d k).eval (n : F) = (n - k + d).choose d := by
+  rw [preHilbert, eval_smul, eval_comp, map_natCast, eval_add, eval_sub, eval_X, eval_natCast,
+    eval_one, smul_eq_mul, ← cast_sub hkn, ← cast_add_one, ← ascPochhammer_eval_cast,
+    ascPochhammer_nat_eq_ascFactorial, ascFactorial_eq_factorial_mul_choose, cast_mul,
+    ← mul_assoc]
+  simp only [isUnit_iff_ne_zero, ne_eq, Ne.symm <| @NeZero.ne' _ _ _ <| @NeZero.charZero _ _
+    ⟨factorial_ne_zero d⟩ .., not_false_eq_true, IsUnit.inv_mul_cancel, one_mul]
+
+variable {F}
+
+/--
+Given `p : F[X]` and `d : ℕ`, the Hilbert polynomial of `p` and `d`. Later we will
+show that `PowerSeries.coeff F n (p * (PowerSeries.invOneSubPow F d))` is equal to
+`(Polynomial.hilbert p d).eval (n : F)` for any large enough `n : ℕ`, which is the
+key property of the Hilbert polynomial.
+-/
+noncomputable def hilbert (p : F[X]) (d : ℕ) : F[X] :=
```

Comment:
I'm not 100% sure that this is the same definition, but here's what I'm proposing:
```lean
/--
Given `p : F[X]` and `d : ℕ`, the Hilbert polynomial of `p` and `d`.
If `p/(1-t)^d` is the Poincaré series of a graded module `M` then
`hilbert p d` is the Hilbert polynomial of this module.
-/
noncomputable def hilbert (p : F[X]) (d : ℕ) : F[X] :=
  ∑ i in p.support,
  (p.coeff i) • preHilbert F (d - 1) i

lemma hilbert_zero (d : ℕ): hilbert (0 : F[X]) d = 0 := by simp [hilbert]

-- is this true?
lemma hilbert_succ (p : F[X]) (d : ℕ) : hilbert (p * (X + 1)) (d + 1) = hilbert p d := sorry
```
and hopefully this definition of `hilbert` is (a)  the same as what you wrote above and (b) easier to work with?

Reply:
> I'm not 100% sure that this is the same definition, but here's what I'm proposing:
> 
> ```lean
> /--
> Given `p : F[X]` and `d : ℕ`, the Hilbert polynomial of `p` and `d`.
> If `p/(1-t)^d` is the Poincaré series of a graded module `M` then
> `hilbert p d` is the Hilbert polynomial of this module.
> -/
> noncomputable def hilbert (p : F[X]) (d : ℕ) : F[X] :=
>   ∑ i in p.support,
>   (p.coeff i) • preHilbert F (d - 1) i
> 
> lemma hilbert_zero (d : ℕ): hilbert (0 : F[X]) d = 0 := by simp [hilbert]
> 
> -- is this true?
> lemma hilbert_succ (p : F[X]) (d : ℕ) : hilbert (p * (X + 1)) (d + 1) = hilbert p d := sorry
> ```
> 
> and hopefully this definition of `hilbert` is (a) the same as what you wrote above and (b) easier to work with?

I have just noticed that this new definition is not correct. Suppose `p` is not divisible by `1 - X` and `d = 0`, then `p * (invOneSubPow F d)` will be a polynomial so we want `hilbert p d` to be the zero polynomial. However, according to our new definition, `hilbert p d = ∑ i in p.support, (p.coeff i) • preHilbert F (d - 1) i = ∑ i in p.support, (p.coeff i) • preHilbert F 0 i = ∑ i in p.support, (p.coeff i) ≠ 0`, where the second equality holds by the definition of the `-` operation on `ℕ`, and the inequality holds because `p` is not divisible by `1`.

So I have come up with the following idea:
```
noncomputable def hilbert (p : F[X]) (d : ℕ) : F[X] :=
  if d = 0 then 0
  else ∑ i in p.support, (p.coeff i) • preHilbert F (d - 1) i
```
This may not look so beautiful as we originally expected, but it is the best idea I can come up with.

What do you think? @erdOne

Reply:
> I'm not 100% sure that this is the same definition, but here's what I'm proposing:
> 
> ```lean
> /--
> Given `p : F[X]` and `d : ℕ`, the Hilbert polynomial of `p` and `d`.
> If `p/(1-t)^d` is the Poincaré series of a graded module `M` then
> `hilbert p d` is the Hilbert polynomial of this module.
> -/
> noncomputable def hilbert (p : F[X]) (d : ℕ) : F[X] :=
>   ∑ i in p.support,
>   (p.coeff i) • preHilbert F (d - 1) i
> 
> lemma hilbert_zero (d : ℕ): hilbert (0 : F[X]) d = 0 := by simp [hilbert]
> 
> -- is this true?
> lemma hilbert_succ (p : F[X]) (d : ℕ) : hilbert (p * (X + 1)) (d + 1) = hilbert p d := sorry
> ```
> 
> and hopefully this definition of `hilbert` is (a) the same as what you wrote above and (b) easier to work with?

Both `hilbert_zero` and `hilbert_succ` turned out to be correct, but the proof of `hilbert_succ` is not so short. As I have decided to include `PowerSeries.coeff_mul_invOneSubPow_eq_hilbert_eval` (i.e. the key property of `Polynomial.hilbert`) in this pull request, I plan to do `hilbert_succ` in some other pull request.